fdbclient/DatabaseContext.cpp

```diff
@@ -0,0 +1,87 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "fdbclient/DatabaseContext.h"
+
+Reference<WatchMetadata> DatabaseContext::getWatchMetadata(int64_t tenantId, KeyRef key) const {
+	const auto it = watchMap.find(std::make_pair(tenantId, key));
+	if (it == watchMap.end())
+		return Reference<WatchMetadata>();
+	return it->second;
+}
+
+void DatabaseContext::setWatchMetadata(Reference<WatchMetadata> metadata) {
+	const WatchMapKey key(metadata->parameters->tenant.tenantId, metadata->parameters->key);
+	watchMap[key] = metadata;
+	// NOTE Here we do *NOT* update/reset the reference count for the key, see the source code in getWatchFuture.
+	// Basically the reference count could be increased, or the same watch is refreshed, or the watch might be cancelled
+}
+
+int32_t DatabaseContext::increaseWatchRefCount(const int64_t tenantID, KeyRef key, const Version& version) {
+	const WatchMapKey mapKey(tenantID, key);
+	watchCounterMap[mapKey].insert(version);
+	return watchCounterMap[mapKey].size();
+}
+
+int32_t DatabaseContext::decreaseWatchRefCount(const int64_t tenantID, KeyRef key, const Version& version) {
+	const WatchMapKey mapKey(tenantID, key);
+	auto mapKeyIter = watchCounterMap.find(mapKey);
+	if (mapKeyIter == std::end(watchCounterMap)) {
+		// Key does not exist. The metadata might be removed by deleteWatchMetadata already.
+		return 0;
+	}
+
+	auto& versionSet = mapKeyIter->second;
+	auto versionIter = versionSet.find(version);
+
+	if (versionIter == std::end(versionSet)) {
+		// Version not found, the watch might be cleared before.
+		return versionSet.size();
+	}
+	versionSet.erase(versionIter);
+
+	const auto count = versionSet.size();
+	// The metadata might be deleted somewhere else, before calling this decreaseWatchRefCount
+	if (auto metadata = getWatchMetadata(tenantID, key); metadata.isValid() && versionSet.size() == 0) {
+		// It is a *must* to cancel the watchFutureSS manually. watchFutureSS waits for watchStorageServerResp, which
+		// holds a reference to the metadata. If the ACTOR is not cancelled, it indirectly holds a Future waiting for
+		// itself.
+		metadata->watchFutureSS.cancel();
+		// It is possible that the DatabaseContext get destructed before those watches that are waiting for the
+		// watchPromise, In this case, the watchPromise will send a broken_promise error which might cause some parts of
+		// the code, e.g. RYWImpl::onError, translates to some other errors that triggers further failures. To ensure
+		// these Future<Void>s fail properly, manually inject a operation_cancelled to them.
```

Comment:
It's not completely clear to me what the sequence of events is here. How is the destruction of `DatabaseContext` factoring into this?

Presumably the issue is that we are somehow decrementing the watch reference count to 0 while someone is still waiting on the watch promise's future. I'm not exactly sure how that happens, it seems like the only waiters come from `watchValueMap` which is the function that seems to be the only caller of `decreaseReferenceCount`. This future is also stored as `WatchMetadata.watchFuture`, but as far as I can tell that field isn't being used and could be deleted.

It's possible this change could work, but it's hard for me to say without really understanding what's happening.

Reply:
The `watchMetadata.watchFuture` is not being used, but on the other hand `watch.watchFuture` is being used and listens to this promise. Any errors forwarded to `watch.watchFuture` will be received by `::watch` and further thrown to corresponding receipts. If it is a `broken_promise`, some of the receipts will escalate the exception. Sending an `operation_cancelled`, which is exactly what happened on the `WatchMetadata` side, would smooth the concern and make the receipts happy.

Reply:
Does this describe the issue and the fix? If so I will update the comment.

Reply:
I'm still a little bit confused, though. If I understand correctly, `watch.watchFuture` is waiting on the promise via `watchValueMap`, and `watchValueMap` seems to be the only place that can actually decrement the reference count. If that were true, though, I would expect that these operations should be ordered correctly (i.e. we stop waiting on the future and then we decrement the reference count), but I don't quite see what I'm missing yet.